_posts/2023-03-09-group-equivariant-cnn-1/group-equivariant-cnn-1.Rmd

```diff
@@ -0,0 +1,148 @@
+date: 2023-03-09
+categories:
+  - Torch
+  - R
+  - Spatial Data
+  - Concepts
+output:
+  distill::distill_article:
+    self_contained: false
+    toc: true
+preview: images/preview.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE, fig.width = 6, fig.height = 6)
+```
+
+This is the first in a series of posts on group-equivariant convolutional neural networks (GCNNs). Today, we keep it short, high-level, and conceptual; examples and implementations will follow. In looking at GCNNs, we are resuming a topic we first [wrote about](https://blogs.rstudio.com/ai/posts/2021-08-26-geometric-deep-learning/) in 2021: [Geometric Deep Learning](https://geometricdeeplearning.com/), a principled, math-driven approach to network design that, since then, has only risen in scope and impact.
+
+## From alchemy to science: Geometric Deep Learning in two minutes
+
+In a nutshell, Geometric Deep Learning is all about deriving network structure from two things: the domain, and the task. The posts will go into a lot of detail, but let me give a quick preview here:
+
+-   By domain, I'm referring to the underlying physical space, and the way it is represented in the input data. For example, images are usually coded as a two-dimensional grid, with values indicating pixel intensities. (At least, that's when the image has a single channel. Otherwise, there are a number of grids, each mapped to a separate channel.)
+-   The task is what we're training the network to do: classification, say, or segmentation. (More accurately, tasks may be different at different stages in the architecture. At each stage, the task in question will have its word to say about how layer design should look.)
+
+For instance, take MNIST. The dataset consists of images of ten digits, 0 to 10, all gray-scale. The task -- unsurprisingly -- is to assign each image the digit represented.
+
+First, consider the domain. A $7$ is a $7$ wherever it appears on the grid. We thus need an operation that is *translation-equivariant*: It flexibly adapts to shifts (translations) in its input. More concretely, in our context, *equivariant* operations are able to detect some object's properties even if that object has been moved, vertically and/or horizontally, to another location. *Convolution*, ubiquitous not just in deep learning, is just such a shift-equivariant operation.
+
+Let me call special attention to the fact that, in equivariance, the essential thing is that "flexible adaptation" (as I casually termed it before). Translation-equivariant operations *do* care about an object's new position; they record a feature not abstractly, but at the object's new position. To see why this is important, consider the network as a whole. When we compose convolutions, we build a hierarchy of feature detectors. That hierarchy should be functional no matter where in the image; in addition, it has to be consistent: Location information needs to be preserved between layers.
```

Comment:
```suggestion
Let me call special attention to the fact that, in equivariance, the essential thing is that "flexible adaptation." Translation-equivariant operations *do* care about an object's new position; they record a feature not abstractly, but at the object's new position. To see why this is important, consider the network as a whole. When we compose convolutions, we build a hierarchy of feature detectors. That hierarchy should be functional no matter where in the image. In addition, it has to be consistent: Location information needs to be preserved between layers.
```

Reply:
1. Since your previous reference was "flexibly adapts," this left me looking for some definition I had missed.
2. Remember to keep your sentence structures simple. If you've got a semicolon and a colon in one, you should probably just break it up with a period.

Reply:
Got it! Think I need to switch full stop and second double quote, though, in the first edited area

Reply:
Oh! Saw another edit of yours, and started to wonder ... Had no idea of this (from https://www.grammarly.com/blog/quotation-marks/) :
 
  When using a period, comma, or exclamation mark with quotation marks, place the punctuation inside the quotes. 
  
  “We won!” shouted the coach to no one in particular. 
  
  If you’re breaking up a sentence, place a comma after the first part, also inside quotes. 
  
  “My goodness,” said my father, “take off your dirty shoes before coming in the house.” 
  
  Just like the use of single quotation marks, however, this applies only to American English. British English places punctuation outside quotation marks.